File: vrouter/src/main/java/com/liminghuang/vrouter/Request.java
package com.liminghuang.vrouter;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;

/**
 * ProjectName: AOP
 * Description: 路由请求.
 * CreateDate: 2020/7/5 8:59 PM
 *
 * @author: <a href="mailto:1569642270@qq.com"">colin</a>
 * @version: 1.0.0
 * @since: 1.0.0
 */
public class Request {
    private IntentCompat mIntentCompat;
    private IDispatcher mDispatcher;
    // TODO: 2020/7/5  

    public static class IntentCompat {
        private Intent intent;
        private int reqCode;
        private Bundle options;

        IntentCompat(Intent intent, int reqCode, Bundle options) {
            this.intent = intent;
            this.reqCode = reqCode;
            this.options = options;
        }
    }

    public static class Builder {
        private Context context;
        private String tag;
        private Bundle params;
        private int flags;
        private Uri uri;
        private int reqCode;
        private Bundle options;

        public Builder setContext(Context context) {
            this.context = context;
            return this;
        }

        public Builder setTag(String tag) {
            this.tag = tag;
            return this;
        }

        public Builder setParams(Bundle params) {
            this.params = params;
            return this;
        }

        public Builder setReqCode(int reqCode) {
            this.reqCode = reqCode;
            return this;
        }

        public Builder setFlags(int flags) {
            this.flags = flags;
            return this;
        }

        public Builder setOptions(Bundle options) {
            this.options = options;
            return this;
        }

        public Builder setUri(Uri uri) {
            this.uri = uri;
            return this;
        }

        public IntentCompat build() {
            Intent intent = null;
            if (tag != null && tag.length() > 0) {
                intent = VRouter.buildIntent(context, tag, params);
            }
            String url = uri.toString();
            if (url != null && url.length() > 0) {
                intent = VRouter.buildIntent(context, url);
            }

            if (intent != null) {
                intent.addFlags(flags);
            }
            return new IntentCompat(intent, reqCode, options);
        }
    }
}

File: vrouter/src/main/kt/com/liminghuang/vrouter/Request.kt
package com.liminghuang.vrouter

import android.content.Context
import android.net.Uri
import android.os.Bundle
import android.util.Log

/**
 * ProjectName: AOP
 * Description: 路由请求.
 * CreateDate: 2020/7/11 8:26 PM
 * @author: <a href="mailto:1569642270@qq.com">colin</a>
 * @version: 1.0.0
 * @since: 1.0.0
 */
class Request(val addressCompat: AddressCompat, val next: Request?) {

    class AddressCompat internal constructor(val context: Context, val address: Address, val reqCode: Int, val options: Bundle?)

    class Builder(private val context: Context) {

        companion object {
            private const val TAG: String = "Request.Builder"
        }

        private var tag: String? = null
        private var params: Bundle? = null
        private var flags = 0
        private var uri: Uri? = null
        private var reqCode = 0
        private var options: Bundle? = null
        private var next: Request? = null

        fun setTag(tag: String?) = apply {
            this.tag = tag
        }

        fun setParams(params: Bundle?) = apply {
            this.params = params
        }

        fun setReqCode(reqCode: Int) = apply {
            this.reqCode = reqCode
        }

        fun setFlags(flags: Int) = apply {
            this.flags = flags
        }

        fun setOptions(options: Bundle?) = apply {
            this.options = options
        }

        fun setUri(uri: Uri?) = apply {
            this.uri = uri
        }

        fun setNext(next: Request?) = apply { this.next = next }

        fun build(): Request? {
            var address: Address? = null
            tag?.let {
                if (it.isNotEmpty()) {
                    Log.d(TAG, "tag: $it");
                    address = VRouter.buildAddress(context, tag!!, params!!)
                }
            }
            val url = uri?.toString()
            url?.let {
                Log.d(TAG, "uri: $it");
                address = VRouter.buildAddress(context, it)
            }
            if (address != null) {
                address!!.intent?.addFlags(flags)
                address!!.intent?.putExtras(params!!)
                // TODO: 2020/7/11
//                 address!!.intent.setData(uri);
            } else {
                Log.d(TAG, "address not found.")
            }

            return address?.let { Request(AddressCompat(context, address!!, reqCode, options), next) }
        }
    }
}